### PR TITLE
N8N-4143 Variables and expressions for pin data nodes without execution

### PR DESCRIPTION
**Pull request by alexgrozav**
https://linear.app/n8n/issue/N8N-4143/make-node-parameter-variables-and-expressions-available-when-using

### CODE REVIEW

**Review by mutdmour — APPROVED** [review]

File: packages/editor-ui/src/components/VariableSelector.vue [hunk 1]
```diff
@@ -491,7 +520,7 @@ export default mixins(
 					}
 				}
 
-				let tempOutputData;
+				let tempOutputData: IVariableSelectorOption[] | null;
```
Comment (ivov) [inline packages/editor-ui/src/components/VariableSelector.vue:523]: This is `undefined` here.

File: packages/editor-ui/src/components/VariableSelector.vue [hunk 2]
```diff
@@ -297,6 +300,32 @@ export default mixins(
...
+				return this.getNodeOutput(nodeName, outputData, filterText, useShort);
+			},
+
+			/**
+			 * Returns the node's output data
+			 *
+			 * @param {string} nodeName The name of the node to get the data of
+			 * @param {INodeExecutionData} outputData The data of the run to get the data of
+			 * @param {string} filterText Filter text for parameters
+			 * @param {boolean} [useShort=false] Use short notation
```
Comment (ivov) [inline packages/editor-ui/src/components/VariableSelector.vue:326]: Nit: I would add a small example here. Not immediately obvious what `short notation` stands for.

File: packages/editor-ui/src/components/VariableSelector.vue [hunk 3]
```diff
@@ -502,7 +531,41 @@ export default mixins(
...
+					let pinDataOptions: IVariableSelectorOption[] = [];
+					parentNode.forEach((parentNodeName) => {
+						const pinData = this.$store.getters['pinDataByNodeName'](parentNodeName);
+
+						if (pinData) {
+							pinDataOptions = pinDataOptions.concat(this.getNodePinDataOutput(parentNodeName, pinData, filterText, true) || []);
+						}
+					});
+
+					if (pinDataOptions.length > 0) {
```
Comment (ivov) [inline packages/editor-ui/src/components/VariableSelector.vue:545]: Out of scope: I find this hard to follow without spending longer, which makes it difficult to properly check the logic. In the future, we should make an effort to refactor this section and ideally this whole 250-line method. Since we are looking to release, now is probably not the time.

**Review by ivov — COMMENTED** [review]
Thanks for taking this on!

1. If we pin a node, connect the next, reference from next to previous using expression, the expression works. If then we add a third node standalone (unconnected), run this third node, and go back to inspect the expression, it now says `[invalid (workflowRunData[parentNode[0]] is undefined)]`

image

2. If we pin a node, connect the next, run the next, when opening the NDV or the expression edit modal, there are Vue store state mutation warnings and watcher callback error warnings in the console. These fire multiple times when selecting any option (pindata-based or not) from the variable selector.

image

File: packages/editor-ui/src/components/VariableSelector.vue (hunk 2 above)
Comment (alexgrozav) [reply packages/editor-ui/src/components/VariableSelector.vue:326]: Added description.

File: packages/editor-ui/src/components/VariableSelector.vue (hunk 3 above)
Comment (alexgrozav) [reply packages/editor-ui/src/components/VariableSelector.vue]: I found it very hard to follow as well. The code there is very complicated and requires deep business-logic understanding. 

I've been doing a lot of trial-and-error work to get to understand what this does and how it interacts with the data.

**Comment by alexgrozav** [comment]
> Thanks for taking this on!
> 
> 1. If we pin a node, connect the next, reference from next to previous using expression, the expression works. If then we add a third node standalone (unconnected), run this third node, and go back to inspect the expression, it now says `[invalid (workflowRunData[parentNode[0]] is undefined)]`
> 
> image

Good catch! Found it as well after more thorough testing. Added a fix for it.

> 2. If we pin a node, connect the next, run the next, when opening the NDV or the expression edit modal, there are Vue store state mutation warnings and watcher callback error warnings in the console. These fire multiple times when selecting any option (pindata-based or not) from the variable selector.
> 
> image

Updated. Rebuilding workflow run data using spreading to be able to manipulate it.

**Review by ivov — APPROVED** [review]
Confirmed, those cases are fixed. FYI pushed a small build fix.

**Review by ivov — COMMENTED** [review]
This does not seem to work:
1. Pin test data to Start, connect to next → connector still grey, no `items` label
2. Pin test data to standalone node, connect to next → connector still grey, no `items` label
3. Pin test data to standalone node, connect to next, keep building chain with every node pinned → all connectors still grey, no `items` label in any node

Reviewed: 7feed12dc

File: packages/editor-ui/src/store.ts [hunk 4]
```diff
@@ -214,6 +215,8 @@ export const store = new Vuex.Store({
 			}
 
 			state.stateIsDirty = true;
+
+			dataPinningEventBus.$emit('pin-data', { [payload.node.name]: payload.data });
```
[the remark this answers is not in the document]
Comment (alexgrozav) [reply packages/editor-ui/src/store.ts:219]: We also need the number of items. See `total: pinData[nodeName].length`.

File: packages/editor-ui/src/views/NodeView.vue [hunk 5]
```diff
@@ -2956,6 +2959,33 @@ export default mixins(
...
+			removePinDataConnections(pinData: PinData) {
+				Object.keys(pinData).forEach((nodeName) => {
+					// @ts-ignore
+					const connections = this.instance.getConnections({
+						source: NODE_NAME_PREFIX + this.getNodeIndex(nodeName),
+					}) as Connection[];
+
+					connections.forEach((connection) => {
+						CanvasHelpers.resetConnection(connection);
+					});
```
[the remark this answers is not in the document]
Comment (alexgrozav) [reply packages/editor-ui/src/views/NodeView.vue]: Updated.

File: packages/editor-ui/src/views/NodeView.vue [hunk 6]
```diff
@@ -3004,10 +3034,19 @@ export default mixins(
...
 
 			this.$externalHooks().run('nodeView.mount');
+
+			dataPinningEventBus.$on('pin-data', (pinData: PinData) => {
+				this.addPinDataConnections(pinData);
+			});
+
+			dataPinningEventBus.$on('unpin-data', (pinData: PinData) => {
+				this.removePinDataConnections(pinData);
+			});
```
[the remark this answers is not in the document]
Comment (alexgrozav) [reply packages/editor-ui/src/views/NodeView.vue]: 1. Done.
2. Because I forgot. 🥲

**Review by ivov — APPROVED** [review]
Amazing work!
